src/main/scala/cromwell/engine/WorkflowDescriptor.scala

```diff
@@ -0,0 +1,196 @@
+    }
+  }
+
+  private def validateWorkflowDescriptor(id: WorkflowId,
+                                         sourceFiles: WorkflowSourceFiles,
+                                         backend: Backend,
+                                         conf: Config): ErrorOr[WorkflowDescriptor] = {
+    val namespace = validateNamespace(sourceFiles.wdlSource, backend)
+    val rawInputs = validateRawInputs(id, sourceFiles.inputsJson)
+    val options = validateWorkflowOptions(id, sourceFiles.workflowOptionsJson, backend)
+    (namespace |@| rawInputs |@| options) { (_, _, _) } match {
+      case scalaz.Success((n, r, o)) =>
+        val validatedDescriptor = for {
+          c <- validateCoercedInputs(id, r, n).disjunction
+          d <- validateDeclarations(id, n, o, c, backend).disjunction
+        } yield WorkflowDescriptor(id, sourceFiles, o, r, n, c, d, backend, configCallCaching(conf), lookupDockerHash(conf))
+        validatedDescriptor.validation
+      case scalaz.Failure(f) => f.list.mkString("").failureNel
+    }
+  }
+
+  private def validateNamespace(source: WdlSource, backend: Backend): ErrorOr[NamespaceWithWorkflow] = {
+    try {
+      NamespaceWithWorkflow.load(source, backend.backendType).successNel
+    } catch {
+      case e: Exception => s"Workflow ${id.toString} unable to load namespace: ${e.getMessage}".failureNel
+    }
+  }
+
+  private def validateWorkflowOptions(id: WorkflowId,
+                                      optionsJson: WorkflowOptionsJson,
+                                      backend: Backend): ErrorOr[WorkflowOptions] = {
+    WorkflowOptions.fromJsonString(optionsJson) match {
+      case Success(o) => validateBackendOptions(id, o, backend)
+      case Failure(e) => s"Workflow ${id.toString} contains bad options JSON: ${e.getMessage}".failureNel
+    }
+  }
+
+  private def validateRawInputs(id: WorkflowId, json: WdlJson): ErrorOr[Map[String, JsValue]] = {
+    Try(json.parseJson) match {
+      case Success(JsObject(inputs)) => inputs.successNel
+      case _ => s"Workflow ${id.toString} contains bad inputs JSON: $json".failureNel
+    }
+  }
+
+  private def validateCoercedInputs(id: WorkflowId,
+                                    rawInputs: Map[String, JsValue],
+                                    namespace: NamespaceWithWorkflow): ErrorOr[WorkflowCoercedInputs] = {
+    namespace.coerceRawInputs(rawInputs) match {
+      case Success(r) => r.successNel
+      case Failure(e) => e.getMessage.failureNel
+    }
+  }
+
+  private def validateBackendOptions(id: WorkflowId, options: WorkflowOptions, backend: Backend): ErrorOr[WorkflowOptions] = {
+    try {
+      backend.assertWorkflowOptions(options)
+      options.successNel
+    } catch {
+      case e: Exception =>
+        s"Workflow ${id.toString} has invalid options for backend ${backend.backendType}: ${e.getMessage}".failureNel
```

Comment:
While I believe you're correct, I'll point out WorkflowId.toString is used heavily throughout. I suspect perhaps at one point it mattered